SDK Sprint 1

jorgeavaldez opened the pull request:
# ML SDK Sprint 1 Requirements

Here are the requirements that have been completed.
## WebService Actionables
- [x] WSD string constructor
- [x] Delete
- [x] Update
- [x] Expose appropriate definition properties according to list
## WebService management client
- [x] Deploy
- [x] Get WSD from experiment
- [x] Get WSD from resource group
- [x] List web services from subscription
- [x] List web services from resource group

Commit:
added properties to new web services class and working on constructor

Commit:
working on auth, finishing the management client

Commit:
added auth and working on management client for code review

Commit:
added the rest of the web service functions and management client functions for this sprint

jorgeavaldez commented:
Compile issues with .NET Core and .NET Framework 4.5. Waiting to test on machine with .NET Framework 4.5 with VM.

Commit:
added get web service from resource group, added example file

Commit:
added list methods and updated example code

jorgeavaldez commented:
Added list methods, missing adding the resource group to web services when pulled from subscription.

Commit:
updated example to include web service lists

Commit:
added parameter for force check under deploy.

Commit:
commented during meeting

jorgeavaldez committed:
fixed most of the issues from review meeting

jorgeavaldez committed:
edited build parameter to include in test project

jorgeavaldez committed:
moved params

jorgeavaldez committed:
fixed setter methods

jorgeavaldez committed:
Merge branch 'azureml.webservice_management' of https://github.com/jorgeavaldez/azure-sdk-for-net into azureml.webservice_management

jorgeavaldez committed:
Update now does a patch, subscription id is now accessable

NonStatic2014 commented on src/ResourceManagement/MachineLearning/Microsoft.Azure.Management.MachineLearning/Generated/Studio.WebService/IWebServiceDefinitionOperations.cs:
```diff
@@ -0,0 +1,44 @@
+    using Models;
+
+    /// <summary>
+    /// WebServiceDefinitionOperations operations.
```
WebServiceDefinitionOperations --> IWebServiceDefinitionOperations

NonStatic2014 commented on src/ResourceManagement/MachineLearning/Microsoft.Azure.Management.MachineLearning/Generated/Studio.WebService/IWebServiceDefinitionOperations.cs:
```diff
@@ -0,0 +1,44 @@
+        /// </param>
+        /// <param name='cancellationToken'>
+        /// The cancellation token.
+        /// </param>
```
Missing <return>

NonStatic2014 commented on src/ResourceManagement/MachineLearning/Microsoft.Azure.Management.MachineLearning/Generated/Studio.WebService/IWebServiceDefinitionOperations.cs:
```diff
@@ -0,0 +1,44 @@
+// Licensed under the MIT License. See License.txt in the project root for
+// license information.
+// 
+// Code generated by Microsoft (R) AutoRest Code Generator 0.16.0.0
```
This is not valid here.

NonStatic2014 commented on src/ResourceManagement/MachineLearning/Microsoft.Azure.Management.MachineLearning/Generated/Studio.WebService/Models/AssetItem.cs:
```diff
@@ -0,0 +1,116 @@
+        /// <summary>
+        /// Initializes a new instance of the AssetItem class.
+        /// </summary>
+        public AssetItem(string name, string type, AssetLocation locationInfo, string id = default(string), IDictionary<string, InputPort> inputPorts = default(IDictionary<string, InputPort>), IDictionary<string, OutputPort> outputPorts = default(IDictionary<string, OutputPort>), IDictionary<string, string> metadata = default(IDictionary<string, string>), IList<ModuleAssetParameter> parameters =  ...
```
nit, please put each parameter in one line. Otherwise, this line is too long.

NonStatic2014 commented on src/ResourceManagement/MachineLearning/Microsoft.Azure.Management.MachineLearning/Generated/Studio.WebService/Models/AssetItem.cs:
```diff
@@ -0,0 +1,116 @@
+        /// Asset's type. Possible values include: 'Module', 'Resource'
+        /// </summary>
+        [JsonProperty(PropertyName = "type")]
+        public string Type { get; set; }
```
Can we make it enum?

NonStatic2014 commented on src/ResourceManagement/MachineLearning/Microsoft.Azure.Management.MachineLearning/Management/IStudioWebService.cs:
```diff
@@ -0,0 +1,79 @@
+    /// Subscription&lt;/li&gt;&lt;li&gt;Get Web Services
+    /// Keys&lt;/li&gt;&lt;/ul&gt;
+    /// </summary>
+    public partial interface IStudioWebService : IDisposable
```
Why "partial"?

NonStatic2014 commented on src/ResourceManagement/MachineLearning/Microsoft.Azure.Management.MachineLearning/Management/IStudioWebService.cs:
```diff
@@ -0,0 +1,79 @@
+namespace Microsoft.Azure.Management.MachineLearning.Studio.WebService
+{
+    using System;
+    using System.Collections.Generic;
```
Please remove dependencies not in use.

NonStatic2014 commented on src/ResourceManagement/MachineLearning/Microsoft.Azure.Management.MachineLearning/Management/IStudioWebService.cs:
```diff
@@ -0,0 +1,79 @@
+        /// generated and included in each request. Default is true.
+        /// </summary>
+        bool? GenerateClientRequestId { get; set; }
+
```
Please remove multiple empty lines.

NonStatic2014 commented on src/ResourceManagement/MachineLearning/Microsoft.Azure.Management.MachineLearning/Management/StudioServiceClient.cs:
```diff
@@ -0,0 +1,243 @@
+﻿using System;
```
Missing "copyright"

NonStatic2014 commented on src/ResourceManagement/MachineLearning/Microsoft.Azure.Management.MachineLearning/Management/StudioServiceClient.cs:
```diff
@@ -0,0 +1,243 @@
+            this._client = new StudioWebService(credentials);            
+        }
+
+        // GetWithHttpMessagesAsync
```
Unexpected

NonStatic2014 commented on src/ResourceManagement/MachineLearning/Microsoft.Azure.Management.MachineLearning/Management/StudioServiceClient.cs:
```diff
@@ -0,0 +1,243 @@
+        /// Gets a web service definition
+        /// </summary>
+        /// Retrive the web service definition from an experiement.
+        /// <param name='operations'>
```
This parameter does not exist.

NonStatic2014 commented on src/ResourceManagement/MachineLearning/Microsoft.Azure.Management.MachineLearning/Management/StudioServiceClient.cs:
```diff
@@ -0,0 +1,243 @@
+        /// Gets a web service definition
+        /// </summary>
+        /// Retrive the web service definition from an experiement.
+        /// <param name='operations'>
```
Same here. Not existing.

NonStatic2014 commented on src/ResourceManagement/MachineLearning/Microsoft.Azure.Management.MachineLearning/Management/StudioServiceClient.cs:
```diff
@@ -0,0 +1,243 @@
+        /// <param name='cancellationToken'>
+        /// The cancellation token.
+        /// </param>
+        public async Task<Microsoft.Azure.Management.MachineLearning.WebServices.Models.WebService> GetAsync(string workspaceId, string experimentId, CancellationToken cancellationToken = default(CancellationToken))
```
The method name seems too generic to me. Please update it with a specific one, e.g. GetWebServiceDefinitionAsync(). Same comments to other "Get" methods.

NonStatic2014 commented on src/ResourceManagement/MachineLearning/Microsoft.Azure.Management.MachineLearning/Management/WebService.cs:
```diff
@@ -0,0 +1,273 @@
+﻿using System;
```
Missing "copyright".

NonStatic2014 commented on src/ResourceManagement/MachineLearning/Microsoft.Azure.Management.MachineLearning/Management/WebService.cs:
```diff
@@ -0,0 +1,273 @@
+
+namespace Microsoft.Azure.MachineLearning
+{
+
```
Please remove unnecessary empty lines and spaces for files under this folder.

NonStatic2014 commented on src/ResourceManagement/MachineLearning/Microsoft.Azure.Management.MachineLearning/Management/WebService.cs:
```diff
@@ -0,0 +1,273 @@
+        /// Internal autogenerated model that holds properties of a web service.
+        /// TODO: Call it webservicedefinition in the swagger
+        /// </summary>
+        internal Management.MachineLearning.WebServices.Models.WebService Definition { get; set; }
```
internal get; private set;

jorgeavaldez replied:
Gets set by the management client

NonStatic2014 commented on src/ResourceManagement/MachineLearning/Microsoft.Azure.Management.MachineLearning/Management/WebService.cs:
```diff
@@ -0,0 +1,273 @@
+            set
+            {
+                this.Definition.Properties.Title = value;
+                this.Definition.Name = value;
```
Expected?

NonStatic2014 commented on src/ResourceManagement/MachineLearning/Microsoft.Azure.Management.MachineLearning/Management/WebService.cs:
```diff
@@ -0,0 +1,273 @@
+
+                    return val;
+                }
+                catch (Exception)
```
catch exception in a getter seems weird. Can you just return Properties.CreatedOn? Its type is also "DateTime?".

Same comments to other catch-exception in getter/setter.

NonStatic2014 commented on src/ResourceManagement/MachineLearning/Microsoft.Azure.Management.MachineLearning/Generated/Studio.WebService/Models/AssetItem.cs:
```diff
@@ -0,0 +1,116 @@
+// Changes may cause incorrect behavior and will be lost if the code is
+// regenerated.
+
+namespace Microsoft.Azure.Management.MachineLearning.Studio.WebService.Models
```
Can you reuse "Microsoft.Azure.Management.MachineLearning.WebService.Models" other than dupe the code?

NonStatic2014 commented on src/ResourceManagement/MachineLearning/Microsoft.Azure.Management.MachineLearning/Management/WebService.cs:
```diff
@@ -0,0 +1,273 @@
+            {
+                if (this.Definition.Properties.ProvisioningState.Equals("Unknown"))
+                {
+                    throw new InvalidOperationException();
```
Throw exception from getter is also uncommon. Since "unknown" is a valid state, I think you could just return it.

Here comes words from MSDN (https://msdn.microsoft.com/en-us/library/ms229006.aspx):

> X AVOID throwing exceptions from property getters.
> Property getters should be simple operations and should not have any preconditions. If a getter can throw an exception, it should probably be redesigned to be a method. Notice that this rule does not apply to indexers, where we do expect exceptions as a result of validating the arguments.

NonStatic2014 commented on src/ResourceManagement/MachineLearning/Microsoft.Azure.Management.MachineLearning/Management/WebService.cs:
```diff
@@ -0,0 +1,273 @@
+        public WebServiceKeys Keys
+        {
+            get { return this.Definition.Properties.Keys; }
+            internal set { this.Definition.Properties.Keys = value; }
```
Since the class is "internal", this is not necessary.

NonStatic2014 commented on src/ResourceManagement/MachineLearning/Microsoft.Azure.Management.MachineLearning/Management/WebService.cs:
```diff
@@ -0,0 +1,273 @@
+        public bool ReadOnlyProperty
+        {
+            get { return this.Definition.Properties.ReadOnlyProperty.Value; }
+            set { this.Definition.Properties.ReadOnlyProperty = value; }
```
here you can check if it is already true, don't set it to false.

NonStatic2014 commented on src/ResourceManagement/MachineLearning/Microsoft.Azure.Management.MachineLearning/Management/WebService.cs:
```diff
@@ -0,0 +1,273 @@
+        public MachineLearningWorkspace MachineLearningWorkspaceLearningWorkspace
+        {
+            get { return this.Definition.Properties.MachineLearningWorkspace; }
+            private set { throw new InvalidOperationException("Tried to set WebService.MachineLearningWorkspace. Cannot set value."); }
```
If you don't want caller set it, just don't provide setter.

NonStatic2014 commented on src/ResourceManagement/MachineLearning/Microsoft.Azure.Management.MachineLearning/Management/WebService.cs:
```diff
@@ -0,0 +1,273 @@
+        /// <summary>
+        /// Constructor for a WebService that takes a Microsoft.Azure.Management.MachineLearning.WebService for the internal Definition.
+        /// </summary>
+        /// <param name="definitionWebService"></param>
```
Missing description.

NonStatic2014 commented on src/ResourceManagement/MachineLearning/Microsoft.Azure.Management.MachineLearning/Management/WebService.cs:
```diff
@@ -0,0 +1,273 @@
+        {
+            string wsDefinition;
+
+            //StreamReader r = new StreamReader("file.txt"); System.IO.
```
Remove it.

NonStatic2014 commented on src/ResourceManagement/MachineLearning/Microsoft.Azure.Management.MachineLearning/Management/WebService.cs:
```diff
@@ -0,0 +1,273 @@
+        /// <summary>
+        /// Reads in a web service from a web service definition json file.
+        /// </summary>
+        /// <param name="webServiceParameters">Parameters to replace within the file if interpolation parameters exist.</param>
```
As we discussed offline, this is not necessary. Please remove it.

NonStatic2014 commented on src/ResourceManagement/MachineLearning/Microsoft.Azure.Management.MachineLearning/Management/WebService.cs:
```diff
@@ -0,0 +1,273 @@
+        /// Updates this web service and updates the internal state. Takes an AutoRest web service model.
+        /// </summary>
+        /// <param name="otherWebService">The other AutoRest web service model from which this web service is being updated.</param>
+        internal void Update(Microsoft.Azure.Management.MachineLearning.WebServices.Models.WebService otherWebService)
```
I think our goal is to avoid end users access classes under "Microsoft.Azure.Management.MachineLearning.WebServices.Models". I think here you want to have is "Microsoft.Azure.MachineLearning.WebService".

NonStatic2014 commented on src/ResourceManagement/MachineLearning/Microsoft.Azure.Management.MachineLearning/Management/WebService.cs:
```diff
@@ -0,0 +1,273 @@
+        /// <summary>
+        /// Synchronizes the internal state keys with the keys on the server side.
+        /// </summary>
+        public void SyncKeys()
```
From the implementation, you sync "down" the key. So you probably want to make the name of method clear "SyncKeysFromCloud" or something like this.

NonStatic2014 commented on src/ResourceManagement/MachineLearning/Microsoft.Azure.Management.MachineLearning/Management/WebServiceManagementClient.cs:
```diff
@@ -0,0 +1,227 @@
+﻿using System;
```
Missing "copyright".

NonStatic2014 commented on src/ResourceManagement/MachineLearning/Microsoft.Azure.Management.MachineLearning/Management/WebServiceManagementClient.cs:
```diff
@@ -0,0 +1,227 @@
+        /// </summary>
+        /// <param name="webServiceDefintionFilePath">The file path for the web service definition file. Expecting a JSON string within.</param>
+        /// <param name="resourceGroup">The resource group in which the web service is deployed.</param>
+        /// <returns></returns>
```
Missing the description of returns.

NonStatic2014 commented on src/ResourceManagement/MachineLearning/Microsoft.Azure.Management.MachineLearning/Management/WebServiceManagementClient.cs:
```diff
@@ -0,0 +1,227 @@
+            Microsoft.Azure.Management.MachineLearning.WebServices.Models.WebService webService,
+            string resourceGroup)
+        {
+            var nWebService = new WebService(webService, resourceGroup,this._client);
```
Hungarian is not allowed.

NonStatic2014 commented on src/ResourceManagement/MachineLearning/Microsoft.Azure.Management.MachineLearning/Management/WebServiceManagementClient.cs:
```diff
@@ -0,0 +1,227 @@
+                        webService.ResourceGroupName + " with the name " + webService.Title + " already exists.");
+                }
+
+                catch (Exception)
```
Catch general exception is not a good idea. Please use specific exception type here.

NonStatic2014 commented on src/ResourceManagement/MachineLearning/Microsoft.Azure.Management.MachineLearning/Management/WebServiceManagementClient.cs:
```diff
@@ -0,0 +1,227 @@
+                    // If this throws an error, then it doesn't exist.
+                    var existingWebService = this._client.WebServices.Get(webService.ResourceGroupName, webService.Title);
+
+                    throw new InvalidOperationException("force flag was set to false; a web service in the resource group " + 
```
Why always throw exception?

NonStatic2014 commented on src/ResourceManagement/MachineLearning/Microsoft.Azure.Management.MachineLearning/Management/WebServiceManagementClient.cs:
```diff
@@ -0,0 +1,227 @@
+        /// </summary>
+        /// <param name="webService">The web service to be deployed.</param>
+        /// <param name="force">Optional: If a web service with the same name exists, just go ahead and override it if true. Otherwise, throw an exception.</param>
+        public void DeployWebService(WebService webService, bool force = false)
```
force -> overwrite
Is it clearer to the user?

NonStatic2014 commented on src/ResourceManagement/MachineLearning/Microsoft.Azure.Management.MachineLearning/Management/WebServiceManagementClient.cs:
```diff
@@ -0,0 +1,227 @@
+        /// <param name="resourceGroupName">The resource group from which to get the web service.</param>
+        /// <param name="webServiceName">The name of the web service to get.</param>
+        /// <returns>The web service.</returns>
+        public WebService GetWebServiceFromResourceGroup(string resourceGroupName, string webServiceName)
```
This actually get from resource group by name. What if we just call it GetWebService()?

NonStatic2014 commented on src/ResourceManagement/MachineLearning/Microsoft.Azure.Management.MachineLearning/Management/WebServiceManagementClient.cs:
```diff
@@ -0,0 +1,227 @@
+                studioDirectWebService.Type,
+                studioDirectWebService.Tags);
+
+            var actualWebService = new WebService(editedStudioDirectWebService, "", this._client);
```
"" -> string.Empty

NonStatic2014 commented on src/ResourceManagement/MachineLearning/Microsoft.Azure.Management.MachineLearning/Management/WebServiceManagementClient.cs:
```diff
@@ -0,0 +1,227 @@
+        /// <param name="force">Optional: If a web service with the same name exists, just go ahead and override it if true. Otherwise, throw an exception.</param>
+        public void DeployWebService(WebService webService, bool force = false)
+        {
+            if (!force)
```
you could do some validation here, e.g. if ResourceGroup is empty or not? I'm asking this because I notice that when you generate a webservice from experiement, the resource group is empty. Other than fail the deployment later, you could catch it here. 

A better solution would be writing a method in WebService class called Validation(). Throw exception when something is wrong or missing. Then call it whereever you need it.

NonStatic2014 commented on src/ResourceManagement/MachineLearning/Microsoft.Azure.Management.MachineLearning/project.json:
```diff
@@ -13,17 +13,21 @@
-    "delaySign": true,
-    "publicSign": false,
+    "delaySign": false,
+    "publicSign": true,
```
Why you need to change them?

NonStatic2014 commented on src/ResourceManagement/MachineLearning/Microsoft.Azure.Management.MachineLearning/Management/WebServiceManagementClient.cs:
```diff
@@ -0,0 +1,227 @@
+            var webServices = this._client.WebServices.List().Value;
+            var webServiceObjects = new List<WebService>();
+
+            foreach (Microsoft.Azure.Management.MachineLearning.WebServices.Models.WebService ws in webServices)
```
Microsoft.Azure.Management.MachineLearning.WebServices.Models.WebService -> var

NonStatic2014 commented on src/ResourceManagement/MachineLearning/Microsoft.Azure.Management.MachineLearning/Management/WebServiceManagementClient.cs:
```diff
@@ -0,0 +1,227 @@
+
+            foreach (Microsoft.Azure.Management.MachineLearning.WebServices.Models.WebService ws in webServices)
+            {
+                // TODO: Get resource group from web service ID.
```
Address all "TODO" before pull request got merged.

NonStatic2014 commented on src/ResourceManagement/MachineLearning/Microsoft.Azure.Management.MachineLearning/project.json:
```diff
@@ -13,17 +13,21 @@
   "dependencies": {
-    "Microsoft.Rest.ClientRuntime.Azure": "[3.3.0,4.0.0)"
+    "Microsoft.Rest.ClientRuntime.Azure": "[3.3.0,4.0.0)",
+    "Microsoft.Rest.ClientRuntime.Azure.Authentication": "2.2.4-preview",
```
Is it necessary?

NonStatic2014 commented on src/ResourceManagement/MachineLearning/Microsoft.Azure.Management.MachineLearning/project.json:
```diff
@@ -13,17 +13,21 @@
   "frameworks": {
     "net45": {
+      "dependencies": {
+        "Microsoft.IdentityModel.Clients.ActiveDirectory": "2.18.206251556"
```
Is it necessary?

jorgeavaldez committed:
Fixed Klein's comments.
Added parameter to GetWebServiceFromExperiment.
Renamed GetWebServiceFromResourceGroup to GetWebService.
Formatting and added license.
Still need to remove the other autogenerated studio code.

jorgeavaldez committed:
fixed serialization settings for the studio api

jorgeavaldez commented:
Resolved the comments above, and fixed the client code for the Studio API

jorgeavaldez committed:
changed webservice constructors to internal

jorgeavaldez committed:
updated studio and deploy

NonStatic2014 commented on src/ResourceManagement/MachineLearning/Microsoft.Azure.Management.MachineLearning/Management/WebServiceManagementClient.cs:
```diff
@@ -155,13 +176,11 @@ public WebService GetWebService(string resourceGroupName, string webServiceName)
             var studioDirectWebService = StudioWebServiceClient.GetWebServiceDefinition(workspaceId, experimentId, workspaceAuthorizationToken);
 
-            Microsoft.Azure.Management.MachineLearning.WebServices.Models.WebServiceProperties editedStudioDirectWebServiceProperties = new WebServiceProperties(
+            Microsoft.Azure.Management.MachineLearning.WebServices.Models.WebServicePropertiesForGraph editedStudioDirectWebServiceProperties = new WebServicePropertiesForGraph(
```
Microsoft.Azure.Management.MachineLearning.WebServices.Models.WebServicePropertiesForGraph -> var

NonStatic2014 commented on src/ResourceManagement/MachineLearning/Microsoft.Azure.Management.MachineLearning/Management/WebServiceManagementClient.cs:
```diff
@@ -90,6 +90,27 @@ public WebService CreateWebServiceObject(string webServiceDefintionFilePath, str
+                string.IsNullOrEmpty(webService.StorageAccount.Key))
+            {
+                throw new InvalidOperationException("The web service you're trying to deploy does not have a valid StorageAccount name/key, you must set one to deploy.");
+            }
```
You could put above validation into a Validation method of the WebService class and invoke it here.

jorgeavaldez replied:
WebService.ValidateForDeploy()

NonStatic2014 commented on src/ResourceManagement/MachineLearning/Microsoft.Azure.Management.MachineLearning/Generated/WebServices/WebServicesOperationsExtensions.cs:
```diff
@@ -227,7 +227,7 @@ public static WebService BeginPatch(this IWebServicesOperations operations, WebS
             /// <param name='operations'>
             /// The operations group for this extension method.
-            /// </param>
+            /// </param>Location
```
What's the "Location" for?

NonStatic2014 commented on src/ResourceManagement/MachineLearning/Microsoft.Azure.Management.MachineLearning/Management/WebService.cs:
```diff
@@ -265,8 +218,15 @@ public void Delete()
         /// Synchronizes the internal state keys with the keys on the server side.
         /// </summary>
-        public void SyncKeys()
+        public void SyncKeysWithCloud()
```
I would recommend you explicitly say "up" or "down" in the method name.

NonStatic2014 commented on src/ResourceManagement/MachineLearning/Microsoft.Azure.Management.MachineLearning/Management/StudioDeserializationWebService.cs:
```diff
@@ -0,0 +1,93 @@
+    /// <summary>
+    /// The purpose of this class is to act as a deserialization target from the Studio API to a WebService object.
+    /// </summary>
+    internal class StudioDeserializationWebService
```
Do you still need this?

jorgeavaldez replied:
Removed

jorgeavaldez committed:
removed extraneous autogenerated studio code and refactored to remove dependency on this. small fixes to webservice'

jorgeavaldez committed:
removed typo
